tests/testsuite/open_namespaces.rs

```diff
@@ -0,0 +1,357 @@
+use cargo_test_support::project;
+use cargo_test_support::registry::RegistryBuilder;
+
+#[cargo_test]
+fn within_namespace_requires_feature() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo::bar"
+                version = "0.0.1"
+                edition = "2015"
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    p.cargo("read-manifest")
+        .masquerade_as_nightly_cargo(&["open-namespaces"])
+        .with_status(101)
+        .with_stderr(
+            r#"error: failed to parse manifest at `[CWD]/Cargo.toml`
+
+Caused by:
+  feature `open-namespaces` is required
+
+  The package requires the Cargo feature called `open-namespaces`, but that feature is not stabilized in this version of Cargo ([..]).
+  Consider adding `cargo-features = ["open-namespaces"]` to the top of Cargo.toml (above the [package] table) to tell Cargo you are opting in to use this unstable feature.
+  See https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#open-namespaces for more information about the status of this feature.
+"#,
+        )
+        .run()
+}
+
+#[cargo_test]
+fn implicit_lib_within_namespace() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                cargo-features = ["open-namespaces"]
+
+                [package]
+                name = "foo::bar"
+                version = "0.0.1"
+                edition = "2015"
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    p.cargo("read-manifest")
```

Comment:
Why chose `cargo read-manifest` over `cargo metadata`?

Reply:
Less output to deal with.

`read-manfiest` is a short term command.  I expect to replace this with `check` as we work to get compilation working.